extensions/parquet/table/src/main/java/io/deephaven/parquet/table/transfer/VariableWidthTransfer.java

```diff
@@ -0,0 +1,135 @@
+        this.context = columnSource.makeGetContext(maxValuesPerPage);
+        this.currentChunkIdx = 0;
+        this.buffer = buffer;
+    }
+
+    @Override
+    public final B getBuffer() {
+        return buffer;
+    }
+
+    final public boolean hasMoreDataToBuffer() {
+        // Unread data present either the table or in the chunk
+        return tableRowSetIt.hasMore() || chunk != null;
+    }
+
+    // TODO Add comments
+    class EncodedData {
+        E data;
+        int numValues; // TODO Add comments
+        int numBytes;
+
+        EncodedData(@NotNull final E data, final int numBytes) {
+            this.data = data;
+            this.numBytes = numBytes;
+            this.numValues = 1;
+        }
+
+        EncodedData(@NotNull final E data, final int numValues, final int numBytes) {
+            this(data, numBytes);
+            this.numValues = numValues;
+        }
+    }
+
+    // TODO Add comments about what to be done before and after calling this method
+    final void transferOnePageToBufferHelper() {
+        if (!hasMoreDataToBuffer()) {
+            return;
+        }
+        boolean stop = false;
+        do {
+            if (chunk == null) {
+                // Fetch a chunk of data from the table
+                final RowSequence rs = tableRowSetIt.getNextRowSequenceWithLength(maxValuesPerPage);
+                // noinspection unchecked
+                chunk = (ObjectChunk<T, Values>) columnSource.getChunk(context, rs);
+                currentChunkIdx = 0;
+            }
+            final int chunkSize = chunk.size();
+            while (currentChunkIdx < chunkSize) {
+                final T data = chunk.get(currentChunkIdx);
+                if (data == null) {
+                    if (!addNullToBuffer()) {
+                        stop = true;
+                        break;
+                    }
+                    currentChunkIdx++;
+                    continue;
+                }
+                EncodedData nextEntry;
+                if (cachedValue == null) {
+                    nextEntry = encodeDataForBuffering(data);
+                } else {
+                    nextEntry = cachedValue;
+                    cachedValue = null;
+                }
+                int numBytesBuffered = getNumBytesBuffered();
+                // Always copy the first entry
+                if ((numBytesBuffered != 0 && numBytesBuffered + nextEntry.numBytes > targetPageSize) ||
+                        !addEncodedDataToBuffer(nextEntry)) {
+                    stop = true;
+                    cachedValue = nextEntry;
+                    break;
+                }
+                currentChunkIdx++;
+            }
+            if (currentChunkIdx == chunk.size()) {
+                chunk = null;
+            }
+        } while (!stop && tableRowSetIt.hasMore());
+    }
+
+    // TODO Add comments explaining what each does
+    abstract boolean addNullToBuffer();
```

Comment:
False would mean that we could not add null to the buffer because its full) and true means we did.
We do have different handling logics for null for different types. For example, for dict encoded strings, we store `NULL_INT` in the IntBuffer, for plain encoded strings, we add `null` to the binary array, and for all vector and array types, we add NULL_INT in the repeatCounts buffer.